tests/wpt/tests/tools/wptrunner/wptrunner/browsers/servodriver.py

```diff
@@ -126,9 +129,45 @@ def make_command(self):
         return [self.webdriver_binary, f"--webdriver={self.port}"] + self.webdriver_args
 
     def cleanup(self):
-        super().cleanup()
         os.remove(self.hosts_path)
 
+    def is_alive(self):
+        if not super().is_alive():
+            return False
+        try:
+            requests.get(f"http://{self.host}:{self.port}/status", timeout=5)    
+        except Exception:
+            self.logger.debug("Servo has shut down normally.")
+            return False
+
+        return True
+
+    def stop(self, force=False):
+        retry_cnt = 0
+        while self.is_alive():
+            self.logger.info("Trying to shut down gracefully by extension command")
+            try:
+                requests.delete(
+                    f"http://{self.host}:{self.port}/session/dummy-session-id/servo/shutdown",
+                    timeout=5
+                )
+            except requests.exceptions.ConnectionError:
+                self.logger.debug("Browser already shut down (connection refused)")
+                break
+            except requests.exceptions.RequestException as e:
+                self.logger.debug(f"Request exception: {e}")
+                break
+            except requests.exceptions.Timeout:
+                self.logger.debug("Request timed out")
+                break
+                
+            retry_cnt += 1
+            if retry_cnt >= self.shutdown_retry:
+                self.logger.debug("Max retry exceeded.")
+                break
+            time.sleep(1)
+        super().stop(force=force)
```

Comment:
Maybe we should add a wait here for `super.is_alive()`, so we know when the process really exited? Currently the binary could be still shutting down, and the process force-killed during the normal shutdown.